zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerTest.java

```diff
@@ -817,6 +817,132 @@ public void shouldUseUpdatedUserTaskDataInCompleteTaskListenerJobHeadersAfterTas
     completeJobs(processInstanceKey, listenerType);
   }
 
+  @Test
+  public void
+      shouldNotIncludeEmptyOrClearedUserTaskPropertiesInCompletingListenerHeadersAfterTaskWasUpdated() {
+    // given: a process instance with a user task configured with an initial assignee, candidate
+    // users/groups, due/follow-up dates, and a `completing` listener
+    final long processInstanceKey =
+        createProcessInstance(
+            createProcessWithZeebeUserTask(
+                userTask ->
+                    userTask
+                        .zeebeAssignee("admin")
+                        .zeebeCandidateUsers("user_A, user_B")
+                        .zeebeCandidateGroups("group_A, group_C, group_F")
+                        .zeebeDueDate("2085-09-21T11:22:33+02:00")
+                        .zeebeFollowUpDate("2095-09-21T11:22:33+02:00")
+                        .zeebeTaskListener(listener -> listener.completing().type(listenerType))));
+
+    final var changes =
+        new UserTaskRecord()
+            // Clear candidate groups and users, due date, and follow-up date
+            .setCandidateGroupsList(List.of())
+            .setCandidateUsersList(List.of())
+            .setDueDate("")
+            .setFollowUpDate("")
+            // Update priority
+            .setPriority(1);
+
+    // when: updating the user task with the specified changes
+    ENGINE.userTask().ofInstance(processInstanceKey).update(changes);
+
+    // and: completing the user task
+    final var userTaskCommand = ENGINE.userTask().ofInstance(processInstanceKey).complete();
+
+    // then: validate the headers of the completing listener job triggered after the task update
+    final var activatedListenerJob = activateJob(processInstanceKey, listenerType);
+
+    assertThat(activatedListenerJob.getCustomHeaders())
+        .describedAs(
+            "Headers should include only the configured, updated or automatically set user task properties")
+        .containsOnly(
+            // Assignee remains unchanged
+            entry(Protocol.USER_TASK_ASSIGNEE_HEADER_NAME, "admin"),
+            // Task key is always included
+            entry(Protocol.USER_TASK_KEY_HEADER_NAME, String.valueOf(userTaskCommand.getKey())),
+            // Default action value for the completing operation
+            entry(Protocol.USER_TASK_ACTION_HEADER_NAME, "complete"),
+            // Updated priority
+            entry(Protocol.USER_TASK_PRIORITY_HEADER_NAME, "1"))
+        .describedAs(
+            "Headers should not include not configured or cleared properties such as candidate groups, candidate users, due date, and follow-up date")
+        .doesNotContainKeys(
+            Protocol.USER_TASK_CANDIDATE_GROUPS_HEADER_NAME,
+            Protocol.USER_TASK_CANDIDATE_USERS_HEADER_NAME,
+            Protocol.USER_TASK_DUE_DATE_HEADER_NAME,
+            Protocol.USER_TASK_FOLLOW_UP_DATE_HEADER_NAME);
```

Comment:
❓ I'm wondering whether it makes more sense to add all of them to task listener jobs regardless, but with empty or null values. This could help clarify to the user that they understand that the property's value correctly. What do you think?

Reply:
That's a great point! This is also more efficient (less data). Let's cover it there in #25890